File: 23/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
import java.util.*;

public class Solution {
    public ListNode mergeKLists(ListNode[] lists) {

        if(lists == null || lists.length == 0) {
            return null;
        } else if(lists.length == 1) {
            return lists[0];
        } else {
            ListNode[] newList = mergeLastTwo(lists);
            return mergeKLists(newList);
        }

    }

    public ListNode[] mergeLastTwo(ListNode[] lists) {
        List<ListNode> toLists = new ArrayList<ListNode>(Arrays.asList(lists));
        ListNode lastOne = toLists.get(toLists.size() - 1);
        ListNode lastTwo = toLists.get(toLists.size() - 2);
        ListNode sorted = merge(lastOne,lastTwo);
        toLists.remove(toLists.size() - 1);
        toLists.remove(toLists.size() - 2);
        toLists.add(sorted);
        return (ListNode[])toLists.toArray();
    }


    public ListNode merge(ListNode first, ListNode second) {
        return null;
    }

    public int headValue(ListNode list) {
        return list.val;
    }

    public int tailValue(ListNode list) {
        ListNode temp = list;

        while(temp.next != null)  {
            temp = temp.next;
        }

        return temp.val;
    }

}



class ListNode {
     int val;
     ListNode next;
     ListNode(int x) { val = x; }
 }

File: 23/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        if not lists:
            return None
        if len(lists) == 1:
            return lists[0]

        head = self.mergeLastTwo(lists[0],lists[1])

        if len(lists) > 2:
            for i in range(2,len(lists)):
                head = self.mergeLastTwo(head,lists[i])

        return head



    def mergeLastTwo(self,list1,list2):

        pointer1 = list1
        pointer2 = list2
        head = ListNode(0)
        pointer = head

        while pointer1 and pointer2:
            if pointer1.val <= pointer2.val:
                pointer.next = pointer1
                pointer1 = pointer1.next
            else:
                pointer.next = pointer2
                pointer2 = pointer2.next
            pointer = pointer.next

        if pointer1:
            pointer.next = pointer1

        if pointer2:
            pointer.next = pointer2

        return head.next
